porcupine/_state.py

```diff
@@ -44,6 +46,41 @@ def _log_tkinter_error(
     log.error("Error in tkinter callback", exc_info=(exc, val, tb))
 
 
+class Quit:
+    ...
+
+
+def open_files(files: Iterable[str]) -> None:
+    tabmanager = get_tab_manager()
+    for path_string in files:
+        if path_string == "-":
+            # don't close stdin so it's possible to do this:
+            #
+            #   $ porcu - -
+            #   bla bla bla
+            #   ^D
+            #   bla bla
+            #   ^D
+            tabmanager.file_queue.put(sys.stdin.read())
+        else:
+            tabmanager.file_queue.put(Path(path_string))
+
+    tabmanager.event_generate("<<OpenFilesQueue>>")
+
+
+def listen_for_files():
+    with _ipc.session() as ipc_message_queue:
+        while True:
+            message = ipc_message_queue.get()
+            if message is Quit:
+                break
+            else:
+                try:
+                    open_files([message])
```

Comment:
I don't like how this invokes tkinter from a thread. The usual "pattern" is to repeatedly check every 50ms or 100ms or so with `after_idle()`.

Reply:
Even if it's only a `event_generate`? It doesn't hurt 🥺

Reply:
My understanding is that Tcl calls (which ultimately all of tkinter boils down to) aren't thread-safe. Feel free to correct me if you have read `_tkinter.c` more carefully than I have :)

Reply:
> you have read `_tkinter.c` more carefully than I have :)

Maybe I did :) But the only interesting thing I found out, is that while a `tk.call` will marshal the call to the interpreter thread, `tk.eval` doesn't do that.

Although the only case that would make `event_generate` thread-unsafe that I can think of is a race-condition on the event queue, there is indeed nothing that would make it explicitly thread-safe (and I assume that the event queue is thread-safe if the Tcl of Tkinter is threaded - which it is mostly these days). But for example, while a `label.config()` always throws an error if not called from the main thread ("Calling Tcl from different apartment"), with `event_generate` this never happened.

Reply:
Resolved. It also simplified the code quite a bit.